### PR TITLE
sync podstatus from leaf to root when root node from ready to notready

### LINKED ISSUE #705: update  pod status from  leaf to root
**What would you like to be added**:

**Why is this needed**:

### PR DESCRIPTION
**Pull request by gao12312**
#### What type of PR is this?

/kind feature
#### What does this PR do?

sync podstatus from leaf to root when root node from ready to notready
#### Which issue(s) does this PR fix?

Fixes #705
#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			CreateFunc: func(createEvent event.CreateEvent) bool {
+				return skipFunc(createEvent.Object)
+			},
+			UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+				pod1 := updateEvent.ObjectOld.(*corev1.Pod)
+				pod2 := updateEvent.ObjectNew.(*corev1.Pod)
+				if !skipFunc(updateEvent.ObjectNew) {
+					return false
+				}
+				return !cmp.Equal(pod1.Status, pod2.Status)
```
Comment (duanmengkk) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]:
```suggestion
				return !reflect.equals(pod1.Status, pod2.Status)
```

**Comment by duanmengkk** [comment]
add unit test

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	return controllerruntime.NewControllerManagedBy(mgr).
+		Named(PodSyncControllerName).
+		WithOptions(controller.Options{}).
+		For(&corev1.Pod{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(createEvent event.CreateEvent) bool {
+				return false
+			},
+			UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+				pod1 := updateEvent.ObjectOld.(*corev1.Pod)
+				pod2 := updateEvent.ObjectNew.(*corev1.Pod)
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: pod1 and pod2 is strange

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+				if !skipFunc(updateEvent.ObjectNew) {
+					return false
+				}
+				return !reflect.DeepEqual(pod1.Status, pod2.Status)
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: skipFunc and reflect.DeepEqual can be written together

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		klog.Errorf("get %s error: %v", request.NamespacedName, err)
+		return reconcile.Result{RequeueAfter: utils.DefaultRequeueTime}, nil
+	}
+
+	if !podutils.IsKosmosPod(&rootpod) {
+		klog.V(4).Info("Pod is not create by kosmos tree, ignore")
+		return reconcile.Result{}, nil
+	}
+
+	if !c.GlobalLeafManager.HasNode(rootpod.Spec.NodeName) {
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: In one2cluster mode, the nodename of the rootpod is not in the subset cluster

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go (hunk 3 above)
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: You should only judge the conditions in status, not all the contents of status

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	if err := c.RootClient.Get(ctx, request.NamespacedName, &rootpod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("pods not found, %s", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("get %s error: %v", request.NamespacedName, err)
+		return reconcile.Result{RequeueAfter: utils.DefaultRequeueTime}, nil
+	}
+
+	if !podutils.IsKosmosPod(&rootpod) {
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: It's the same as skipFunc

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	leafPod := &corev1.Pod{}
+	err = lcr.Client.Get(ctx, request.NamespacedName, leafPod)
+	if err != nil {
+		klog.Errorf("Failed to get leaf pod  %v", leafPod.Name)
+		return reconcile.Result{}, nil
+	}
+
+	if podutils.IsKosmosPod(leafPod) && !reflect.DeepEqual(rootpod.Status, leafPod.Status) {
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go:108]: status.condition?

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		rPodCopy := rootpod.DeepCopy()
+		rPodCopy.Status = leafPod.Status
+		podutils.FitObjectMeta(&rPodCopy.ObjectMeta)
+		if err := c.RootClient.Status().Update(ctx, rPodCopy); err != nil && !apierrors.IsNotFound(err) {
+			klog.Errorf("error while updating rootpod status in kubernetes, %s", err)
+			return reconcile.Result{}, nil
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go:114]: requeue?

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil {
+		klog.Errorf("Failed to get leaf pod  %v", leafPod.Name)
+		return reconcile.Result{}, nil
+	}
+
+	if podutils.IsKosmosPod(leafPod) && !reflect.DeepEqual(rootpod.Status, leafPod.Status) {
+		rPodCopy := rootpod.DeepCopy()
+		rPodCopy.Status = leafPod.Status
+		podutils.FitObjectMeta(&rPodCopy.ObjectMeta)
+		if err := c.RootClient.Status().Update(ctx, rPodCopy); err != nil && !apierrors.IsNotFound(err) {
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: apierrors.IsNotFound is redundant

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+		Named(PodSyncControllerName).
+		WithOptions(controller.Options{}).
+		For(&corev1.Pod{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(createEvent event.CreateEvent) bool {
+				return false
+			},
+			UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+				old := updateEvent.ObjectOld.(*corev1.Pod)
+				curr := updateEvent.ObjectNew.(*corev1.Pod)
+				if !skipFunc(updateEvent.ObjectNew) {
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go:52]: old and new

File: pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+		klog.Errorf("Failed to get leaf pod  %v", leafPod.Name)
+		return reconcile.Result{}, nil
+	}
+
+	if podutils.IsKosmosPod(leafPod) && !reflect.DeepEqual(rootpod.Status, leafPod.Status) {
+		rPodCopy := rootpod.DeepCopy()
+		rPodCopy.Status = leafPod.Status
+		podutils.FitObjectMeta(&rPodCopy.ObjectMeta)
+		if err := c.RootClient.Status().Update(ctx, rPodCopy); err != nil {
+			if apierrors.IsNotFound(err) {
```
Comment (rxy0210) [inline pkg/clustertree/cluster-manager/controllers/pod/rootpodstatus_sync_controller.go]: no if else

**Pull request merged**